### PR TITLE
[RELAY][PASS] FoldScaleAxis Forward

### PR DESCRIPTION
**Pull request by tqchen**
This is a first serious attempt to implement an NN related optimization pass on relay. This PR contains the following changes:

- [x] Make squeeze's attrs numpy compatible
- [x] Add option to type resolver to write back checked_type to type_annotation if the original one is None
    - There could be some debate on whether we should do so, but it is in general helpful in NN optimizations, where later phase of optimization will rely on knowing types of all arguments  @jroesch 
- [x] Support a general FoldScaleAxis Forward 
   - This pass assumes that the program is in dataflow form and inlines all lets
   - So far only forward direction is implemented.

Hopefully, this can serve as an example of how optimizations can be done in NNVMv2(relay) IR, and how the infrastructure of relay makes writing optimization in a more principled fashion.

## Goal
Fold the scaling of axis(usually caused by BatchNorm) into weight of conv2d in the future. For example

Old:
```
%1 = multiply(%x, %scale)
%2 = conv2d(%1, %w, data_layout="NHWC")
```
Transformed:
```
# scale weight's input channel
%1 = multiply(%w, expand_dims(%scale, axis=1, num_newaxis=2))
%2 = conv2d(%x, %1, data_layout="NHWC")
```

Further constant folding can fold the multiplication and we remove the scaling in the network.

## The Algorithm
While so far only the forward direction is implemented. The general idea is that we transform Expr to tuple of ```(value, axes, scale)```, where the final result satisfies:
```
result = value
for i, k in enumerate(axes):
   k-ith dimension of result *= i-th dimension of scale
```
Then we can propagate this signal along and fold the scale if necessary.  However, it is possible that certain scale may never be consumed if there is no dense/conv2d that follow multiplication.

In order to make sure all the scale we sent out can be consumed eventually, we run a backward "preparation phase", which propagates the demand of the potential axes scaling back to its input.

The new pass is more general than the FoldScaleAxis in nnvm
- The new pass support arbitrary scaling of multiple axes(although the further implementation is necessary), which could be helpful in NHWCc case,
- The new pass support folding forward into two separate path of conv2d, which was not possible in NNVM.

### CODE REVIEW

**Comment by tqchen** [comment]
@jroesch @ZihengJiang @yzhliu @zhiics @FrozenGene @merrymercy @srkreddy1238 @masahi please review

File: src/relay/pass/fold_scale_axis.cc [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+// Per operator defs for FScaleAxisForward
+//----------------------------------------------
+
+// Intermediate operators
+Array<AxesSet> ReluForwardPrep(const Call& call, AxesSet out) {
+  return {out};
+}
+
+STuple ReluForwardTransform(const Call& ref_call,
+                              const AxesSet& expected_axes,
```
Comment (zhiics) [inline src/relay/pass/fold_scale_axis.cc]: Is there any special reason to pass `expected_axes`? It looks it is unused.

File: src/relay/pass/fold_scale_axis.cc [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    Expr lhs = Divide(sargs[0]->value, scale);
+    rnode->value = CallNode::make(ref_call->op, {lhs, sargs[1]->value},
+                                  ref_call->attrs, ref_call->type_args);
+    rnode->scale = sargs[1]->scale;
+    rnode->axes = sargs[1]->axes;
+  }
+  return STuple(rnode);
+}
+
+RELAY_REGISTER_OP("add")
```
Comment (zhiics) [inline src/relay/pass/fold_scale_axis.cc]: s/add/addition? To keep consistent to `subtract`, `multiply`, and `divide`

File: src/relay/pass/fold_scale_axis.cc [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+      c_small_axis < 0 &&
+      (param->groups == 1 || is_depthwise_conv2d)) {
+    data_axes = {c_big_axis};
+  }
+  return {data_axes, NullValue<AxesSet>()};
+}
+
+// Conv2D consumes the scale axis during transformation.
+STuple Conv2DForwardTransform(const Call& ref_call,
+                                const AxesSet& expected_axes,
```
Comment (zhiics) [inline src/relay/pass/fold_scale_axis.cc]: `expected_axes` is unused. Should it be attached to `rnode`?

File: src/relay/pass/fold_scale_axis.cc [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+    return {out_axes, none};
+  } else if (MatchBroadcastToLeftAxes(trhs, tlhs, out_axes)) {
+    return {none, out_axes};
+  } else {
+    return {none, none};
+  }
+}
+
+STuple AddSubForwardTransform(const Call& ref_call,
+                              const AxesSet& expected_out_axes,
```
Comment (zhiics) [inline src/relay/pass/fold_scale_axis.cc]: `expected_out_axes ` is unused.

File: python/tvm/relay/op/transform.py [hunk 5]
```diff
@@ -49,17 +49,17 @@ def transpose(data, axes=None):
...
     Parameters
     ----------
     data : relay.Expr
         The input data to the operator.
 
-    axes : None or List[int]
+    axis : None or List[int]
         Axes to remove.
-        If axes = [] or = None, remove all axis of dimensions 1.
+        If axes = None, remove all axis of dimensions 1.
```
Comment (ZihengJiang) [inline python/tvm/relay/op/transform.py]: change the comment also

File: include/tvm/relay/expr_functor.h [hunk 6]
```diff
@@ -161,15 +168,16 @@ class ExprMutator
...
    *
    * Can be overloaded to transform the types in arbitrary
    * ways, one way would be to define a sub-class of type
    * visitor for types which transform them appropriately.
    */
   virtual Type VisitType(const Type& t);
 
- private:
+ protected:
   /*! \brief Internal map used for memoization. */
```
[the remark this answers is not in the document]
Comment (tqchen) [reply include/tvm/relay/expr_functor.h]: memoization is a term https://en.wikipedia.org/wiki/Memoization

File: src/relay/pass/fold_scale_axis.cc [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+ * \param lhs The left axis.
+ * \param rhs The right axis.
+ * \return The result of the inersection.
+ */
+AxesSet Intersect(const AxesSet& lhs, const AxesSet& rhs) {
+  if (!lhs.defined()) return lhs;
+  if (!rhs.defined()) return rhs;
+  AxesSet ret;
+  size_t i = 0, j = 0;
+  while (i < lhs.size() && j < rhs.size()) {
```
[the remark this answers is not in the document]
Comment (tqchen) [reply src/relay/pass/fold_scale_axis.cc]: Yes, this is the requirement of axis set, Thanks for pointing this out, will add a comment block about it

File: src/relay/pass/fold_scale_axis.cc (hunk 1 above)
Comment (tqchen) [reply src/relay/pass/fold_scale_axis.cc]: expected_axes is necessary for Multiply, which is used to decide whether it should produce STuple or not(if expected_axes is None, it means there are not conv2d down the path).

File: src/relay/pass/fold_scale_axis.cc (hunk 2 above)
Comment (tqchen) [reply src/relay/pass/fold_scale_axis.cc]: This follow numpy convention

File: include/tvm/relay/op.h [hunk 8]
```diff
@@ -74,16 +74,42 @@ class OpNode : public relay::ExprNode {
     v->Visit("support_level", &support_level);
   }
 
+  /*!
+   * \brief Check that if current op is a "primtive operator".
+   * That is the arguments are all type variables, and there is a single
+   * type relation applied to the input and output types.
+   */
+  bool IsPrimitiveOp() const {
+    if (is_primitive_ != -1) return is_primitive_ != 0;
```
Comment (MarisaKirisame) [inline include/tvm/relay/op.h]: if (is_primitive == -1) is_primitive_ = this->Is...;
return is_primitive_ != 0;

File: src/relay/pass/pattern_util.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  for (size_t i = 0; i < tlhs->shape.size(); ++i) {
+    if (j < lhs_axes.size() && i == static_cast<size_t>(lhs_axes[j]->value)) {
+      if (i < base || !equal(tlhs->shape[i], trhs->shape[i - base])) {
+        return false;
+      }
+      ++j;
+    } else {
+      if (i >= base) {
```
Comment (MarisaKirisame) [inline src/relay/pass/pattern_util.h]: else if

File: python/tvm/relay/op/transform.py (hunk 5 above)
Comment (tqchen) [reply python/tvm/relay/op/transform.py]: do not get what do you mean

File: src/relay/pass/fold_scale_axis.cc [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+  void VisitExpr_(const FunctionNode* op) {
+    ExprVisitor::VisitExpr_(op);
+    auto flazy = [this, op] {
+      this->Update(op->body, NullValue<AxesSet>());
+    };
+    flist_.push_back(flazy);
+  }
+
+  void VisitExpr_(const TupleGetItemNode* op) {
+    // pass, do nothing
```
Comment (MarisaKirisame) [inline src/relay/pass/fold_scale_axis.cc]: why dont you visit inside? maybe there is opt ability inside

**Comment by tqchen** [comment]
Thanks @zhiics @ZihengJiang for helpful reviews, please check again

File: src/relay/pass/fold_scale_axis.cc [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+      for (const Expr& field : op->fields) {
+        this->Update(field, NullValue<AxesSet>());
+      }
+    };
+    flist_.push_back(flazy);
+  }
+
+  void VisitExpr_(const IfNode* op) {
+    ExprVisitor::VisitExpr_(op);
+    // do pass through condition.
```
Comment (MarisaKirisame) [inline src/relay/pass/fold_scale_axis.cc]: remove .

File: src/relay/pass/fold_scale_axis.cc (hunk 3 above)
Comment (MarisaKirisame) [reply src/relay/pass/fold_scale_axis.cc]: also indent

File: src/relay/pass/fold_scale_axis.cc [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+RELAY_REGISTER_OP("nn.leaky_relu")
+.set_attr<FForwardPrep>("FScaleAxisForwardPrep", ReluForwardPrep);
+
+RELAY_REGISTER_OP("nn.leaky_relu")
+.set_attr<FForwardTransform>("FScaleAxisForwardTransform", ReluForwardTransform);
+
+// AddSub
+Array<AxesSet> AddSubForwardPrep(const Call& call, AxesSet out_axes) {
+  const auto* tlhs = call->args[0]->type_as<TensorTypeNode>();
+  const auto* trhs = call->args[1]->type_as<TensorTypeNode>();
```
Comment (ZihengJiang) [inline src/relay/pass/fold_scale_axis.cc]: Do we need to add check here?

Comment (tqchen) [reply src/relay/pass/fold_scale_axis.cc]: type_as does the check

File: python/tvm/relay/op/transform.py (hunk 5 above)
Comment (tqchen) [reply python/tvm/relay/op/transform.py]: gotcha

File: include/tvm/relay/op.h (hunk 8 above)
Comment (tqchen) [reply include/tvm/relay/op.h]: This code path is more frequent(cached) and can likely hint compiler to do so

**Pull request merged**